File: compiler/tokenizer.py
KEYWORDS = ["from", "use", "as", "int", "float", "void", "array", "string", "bool", "if", "else", "for", "in", "set", "return", "typeof", "while", "bool", "true", "false", "func", "class", "new", "this", "static", "virtual", "arg", "fopen", "fclose", "fread", "fwrite", "fseek", "ftell", "malloc", "free"]
SINGLE_CHAR_SYMBOLS = ["(", ")", "[", "]", "{", "}", ",", ".", ":", "!", "+", "-", "*", "/", "=", ">", "<"]
DOUBLE_CHAR_SYMBOLS = ["==", "!=", "+=", ">=", "<=", "&&", "||"]
QUOTES = ['"', "'"]
WHITESPACE = [" ", "\t", "\n"] # space, tab, newline (i think thats a tab char i may be wrong)
class Token:
    def __init__(self, _type, value, line=None, column=None):
        self.type = _type # we use type to stop conflicts with pythons built in type() function
        self.value = value
        self.line = line
        self.column = column
    def __repr__(self):
        return f"{self.type}: {self.value}"
class CompilerError(Exception):
    def __init__(self, message, severity, file_path=None):
        file_info = f" in {file_path}" if file_path and file_path != "unknown" else ""
        super().__init__(message + f"{file_info} (Severity: {severity})")
        # change text colour to yellow for warnings and stop compiling and turn text red for error

class Tokenizer:
    def __init__(self, source: str, file_path=None):
        self.source = source
        self.position = 0
        self.line = 1
        self.column = 1
        self.file_path = file_path or "unknown"
    def current_char(self):
        return self.source[self.position]
    def advance(self, offset=1):
        for _ in range(offset):
            if self.position < len(self.source):
                if self.source[self.position] == '\n':
                    self.line += 1
                    self.column = 1
                else:
                    self.column += 1
            self.position += 1
    def is_at_end(self):
        if self.position >= len(self.source):
            return True
        return False
    def peek_next(self):
        if self.position + 1 < len(self.source):
            return self.source[self.position+1]
        return None
    def skip_whitespace(self):
        while not self.is_at_end() and self.current_char() in WHITESPACE:
            self.advance()
    def skip_comment(self):
        if not self.is_at_end() and self.current_char() == "/" and self.peek_next() == "/":
            # C-style comment: //
            self.advance()  # consume first /
            self.advance()  # consume second /
            while not self.is_at_end() and self.current_char() != "\n":
                self.advance()
    def read_word(self):
        word = ""
        start_line = self.line
        start_column = self.column
        while not self.is_at_end() and (self.current_char().isalnum() or self.current_char() == "_"):
            word += self.current_char()
            self.advance()
        if word in KEYWORDS:
            return Token("KEYWORD", word, start_line, start_column)
        else:
            return Token("NAME", word, start_line, start_column)
    def read_number(self):
        number = ""
        has_decimal = False
        has_exponent = False
        start_line = self.line
        start_column = self.column
        
        # Read digits before decimal point
        while not self.is_at_end() and self.current_char().isdigit():
            number += self.current_char()
            self.advance()
        
        # Handle decimal point
        if not self.is_at_end() and self.current_char() == '.':
            has_decimal = True
            number += '.'
            self.advance()
            
            # Read digits after decimal point (at least one digit required)
            if not self.is_at_end() and self.current_char().isdigit():
                while not self.is_at_end() and self.current_char().isdigit():
                    number += self.current_char()
                    self.advance()
            else:
                raise CompilerError(f"Invalid float literal at line {start_line}: expected digits after decimal point", "ERROR", self.file_path)
        
        # Handle scientific notation
        if not self.is_at_end() and self.current_char().lower() == 'e':
            has_exponent = True
            number += 'e'
            self.advance()
            
            # Handle exponent sign
            if not self.is_at_end() and self.current_char() in '+-':
                number += self.current_char()
                self.advance()
                
            # Read exponent digits (at least one digit required)
            if not self.is_at_end() and self.current_char().isdigit():
                while not self.is_at_end() and self.current_char().isdigit():
                    number += self.current_char()
                    self.advance()
            else:
                raise CompilerError(f"Invalid float literal at line {start_line}: expected digits in exponent", "ERROR", self.file_path)
        
        # Determine token type based on whether we saw a decimal point or exponent
        is_float = has_decimal or has_exponent
        token_type = "FLOAT_LITERAL" if is_float else "INTEGER_LITERAL"
        return Token(token_type, number, start_line, start_column)
    def read_string(self):
        string = ""
        # double check the current char is a quote
        if not self.current_char() in QUOTES:
            return None
        start_line = self.line
        start_column = self.column
        self.advance()
        while not self.is_at_end() and self.current_char() not in QUOTES:
            if self.current_char() == '\\' and not self.is_at_end():
                # Handle escape sequences
                self.advance()
                if self.is_at_end():
                    break
                escape_char = self.current_char()
                if escape_char == 'n':
                    string += '\n'
                elif escape_char == 't':
                    string += '\t'
                elif escape_char == 'r':
                    string += '\r'
                elif escape_char == '\\':
                    string += '\\'
                elif escape_char == '"':
                    string += '"'
                elif escape_char == "'":
                    string += "'"
                elif escape_char == '0':
                    string += '\0'
                else:
                    # Unknown escape, keep as-is
                    string += '\\' + escape_char
            else:
                string += self.current_char()
            self.advance()
        self.advance()
        return Token("STRING", string, start_line, start_column)
    def read_symbol(self):
        if not self.is_at_end():
            # Check for // comment start first
            if self.current_char() == "/" and self.peek_next() == "/":
                # Skip the entire comment
                self.advance()  # consume first /
                self.advance()  # consume second /
                while not self.is_at_end() and self.current_char() != "\n":
                    self.advance()
                # Return None to indicate no token was produced (comment skipped)
                return None
            elif self.peek_next() and (self.current_char() + self.peek_next()) in DOUBLE_CHAR_SYMBOLS:
                token =  Token("SYMBOL", (self.current_char() + self.peek_next()), self.line, self.column)
                self.advance(2)
                return token
            elif self.current_char() in SINGLE_CHAR_SYMBOLS:
                token = Token("SYMBOL", self.current_char(), self.line, self.column)
                self.advance()
                return token
            else:
                raise CompilerError(f"Unknown symbol '{self.current_char()}' at line {self.line}, column {self.column}", "ERROR", self.file_path)
    def tokenize(self):
        tokens = []
        while not self.is_at_end():
            self.skip_whitespace()
            self.skip_comment()
            self.skip_whitespace()
            if self.is_at_end():
                break
            if self.current_char() in QUOTES:
                tokens.append(self.read_string())
            elif self.current_char() == '-':
                # Check if this is a negative number (unary minus)
                # Look ahead to see if next character starts a number
                next_char = self.peek_next() if not self.is_at_end() else None
                
                # If next character is a digit, this is a negative number
                if next_char and next_char.isdigit():
                    self.advance()  # consume the minus
                    number = "-"
                    start_line = self.line
                    start_column = self.column
                    
                    # Read the actual number
                    num_token = self.read_number()
                    number += num_token.value
                    
                    # Determine if it's float or int based on the number token
                    token_type = "FLOAT_LITERAL" if num_token.type == "FLOAT_LITERAL" else "INTEGER_LITERAL"
                    tokens.append(Token(token_type, number, start_line, start_column))
                else:
                    # This is a binary minus operator
                    tokens.append(self.read_symbol())
            elif self.current_char() in SINGLE_CHAR_SYMBOLS or (self.peek_next() and (self.current_char() + self.peek_next()) in DOUBLE_CHAR_SYMBOLS) or (self.current_char() == "/" and self.peek_next() == "/"):
                symbol_token = self.read_symbol()
                if symbol_token is not None:  # Skip None tokens (comments)
                    tokens.append(symbol_token)
            elif self.current_char().isalpha() or self.current_char() == "_":
                tokens.append(self.read_word())
            elif self.current_char().isdigit():
                tokens.append(self.read_number())
            else:
                # Debug: show what character is causing issue
                char = self.current_char() if not self.is_at_end() else "EOF"
                raise CompilerError(f"Tokenizer failed at char {self.position}: '{char}' (line {self.line}, col {self.column})", "ERROR", self.file_path)
        return tokens

File: src/tokenizer.cpp
#include "tokenizer.hpp"
#include <cctype>
#include <iostream>

const std::unordered_set<std::string> KEYWORDS = {"from", "use", "as", "int", "float", "void", "array", "dict", "string", "if", "elif", "else", "for", "in", "set", "return", "typeof", "hasattr", "classof", "while", "bool", "true", "false", "func", "class", "new", "this", "static", "virtual", "null", "try", "catch", "throw", "except", "break", "external", "dynamic"};
const std::unordered_set<std::string> SINGLE_CHAR_SYMBOLS = {"(", ")", "[", "]", "{", "}", ",", ".", ":", "!", "+", "-", "*", "/", "=", ">", "<"};
const std::unordered_set<std::string> DOUBLE_CHAR_SYMBOLS = {"==", "!=", "+=", ">=", "<=", "&&", "||"};
const std::unordered_set<std::string> QUOTES = {"\"", "'"};
const std::unordered_set<std::string> WHITESPACE = {" ", "\t", "\n", "\r"};

Tokenizer::Tokenizer(std::string& s, std::string& fp)
    : source(s), file_path(fp), position(0), line(1), col(1) {}

char Tokenizer::current_char() {
    return source[position];
}

void Tokenizer::advance(int offset) {
    for (int i = 0; i < offset; i++) {
        if (position < (int)source.length()) {
            if (source[position] == '\n') {
                line++;
                col = 1;
            } else {
                col++;
            }
        }
        position++;
    }
}

bool Tokenizer::is_at_end() {
    return position >= (int)source.length();
}

char Tokenizer::peek_next(int offset) {
    if (position + offset < (int)source.length()) {
        return source[position + offset];
    }
    return '\0';
}

void Tokenizer::skip_whitespace() {
    while (!is_at_end() && WHITESPACE.count(std::string(1, current_char()))) {
        advance();
    }
}

void Tokenizer::skip_comment() {
    if (!is_at_end() && current_char() == '/' && peek_next() == '/') {
        advance();
        advance();
        while (!is_at_end() && current_char() != '\n' && current_char() != '\r') {
            advance();
        }
    }
}

Token Tokenizer::read_word() {
    std::string word = "";
    int start_line = line;
    int start_col = col;
    while (!is_at_end() && (std::isalnum(current_char()) || current_char() == '_')) {
        word += current_char();
        advance();
    }
    if (KEYWORDS.count(word)) {
        return Token(T_KEYWORD, word, start_line, start_col);
    } else {
        return Token(T_NAME, word, start_line, start_col);
    }
}

Token Tokenizer::read_number() {
    std::string number = "";
    bool has_decimal = false;
    bool has_exponent = false;
    int start_line = line;
    int start_col = col;

    while (!is_at_end() && std::isdigit(current_char())) {
        number += current_char();
        advance();
    }

    if (!is_at_end() && current_char() == '.') {
        has_decimal = true;
        number += ".";
        advance();
        if (!is_at_end() && std::isdigit(current_char())) {
            while (!is_at_end() && std::isdigit(current_char())) {
                number += current_char();
                advance();
            }
        } else {
            throw CompilerError(
                std::string("Invalid float literal at line ") + std::to_string(start_line) +
                ": expected digits after decimal point", "ERROR", file_path);
        }
    }

    if (!is_at_end() && std::tolower(current_char()) == 'e') {
        has_exponent = true;
        number += "e";
        advance();
        if (!is_at_end() && std::string("+-").find(current_char()) != std::string::npos) {
            number += current_char();
            advance();
        }
        if (!is_at_end() && std::isdigit(current_char())) {
            while (!is_at_end() && std::isdigit(current_char())) {
                number += current_char();
                advance();
            }
        } else {
            throw CompilerError(
                std::string("Invalid float literal at line ") + std::to_string(start_line) +
                ": expected digits in exponent", "ERROR", file_path);
        }
    }

    bool is_float = has_decimal || has_exponent;
    int token_type = is_float ? T_FLOAT_LITERAL : T_INT_LITERAL;
    return Token(token_type, number, start_line, start_col);
}

Token Tokenizer::read_string() {
    std::string str = "";
    char quote_char = current_char();
    int start_line = line;
    int start_col = col;
    advance();

    while (!is_at_end() && current_char() != quote_char) {
        if (current_char() == '\\' && !is_at_end()) {
            advance();
            if (is_at_end()) break;
            char escape_char = current_char();
            switch (escape_char) {
                case 'n':  str += '\n'; advance(); break;
                case 't':  str += '\t'; advance(); break;
                case 'r':  str += '\r'; advance(); break;
                case '\\': str += '\\'; advance(); break;
                case '"':  str += '"';  advance(); break;
                case '\'': str += '\''; advance(); break;
                default:
                    if (escape_char >= '0' && escape_char <= '7' &&
                        peek_next(1) >= '0' && peek_next(1) <= '7' &&
                        peek_next(2) >= '0' && peek_next(2) <= '7') {
                        std::string octal_str;
                        octal_str += escape_char;
                        octal_str += peek_next(1);
                        octal_str += peek_next(2);
                        int octal_val = std::stoi(octal_str, nullptr, 8);
                        str += static_cast<char>(octal_val);
                        advance(3);
                    } else {
                        str += '\\';
                        str += escape_char;
                        advance();
                    }
                    break;
            }
        } else {
            str += current_char();
            advance();
        }
    }

    if (is_at_end()) {
        throw CompilerError(
            std::string("Unterminated string literal at line ") + std::to_string(start_line) +
            ", column " + std::to_string(start_col), "ERROR", file_path);
    }
    advance();
    return Token(T_STRING, str, start_line, start_col);
}

Token Tokenizer::read_symbol() {
    if (!is_at_end()) {
        std::string sym;
        sym += current_char();
        if (peek_next() != '\0') {
            sym += peek_next();
        }
        if (peek_next() != '\0' && DOUBLE_CHAR_SYMBOLS.count(sym)) {
            Token token(T_SYMBOL, sym, line, col);
            advance(2);
            return token;
        } else if (SINGLE_CHAR_SYMBOLS.count(std::string(1, current_char()))) {
            Token token(T_SYMBOL, std::string(1, current_char()), line, col);
            advance();
            return token;
        } else {
            throw CompilerError(
                std::string("Unknown symbol '") + std::string(1, current_char()) +
                "' at line " + std::to_string(line) + ", column " + std::to_string(col),
                "ERROR", file_path);
        }
    }
    throw CompilerError("Unexpected end of input while reading symbol", "ERROR", file_path);
}

std::vector<Token> Tokenizer::tokenize() {
    std::vector<Token> tokens;
    while (!is_at_end()) {
        while (!is_at_end()) {
            std::size_t start_pos = position;
            skip_whitespace();
            skip_comment();
            if (position == static_cast<int>(start_pos)) {
                break;
            }
        }
        if (is_at_end()) break;

        if (QUOTES.count(std::string(1, current_char()))) {
            tokens.push_back(read_string());
        } else if (current_char() == '-') {
            char next_char = !is_at_end() ? peek_next() : '\0';
            if (next_char != '\0' && std::isdigit(next_char)) {
                advance();
                std::string number = "-";
                int start_line = line;
                int start_col = col;
                Token num_token = read_number();
                number += num_token.value;
                int token_type = num_token.type == T_FLOAT_LITERAL ? T_FLOAT_LITERAL : T_INT_LITERAL;
                tokens.push_back(Token(token_type, number, start_line, start_col));
            } else {
                tokens.push_back(read_symbol());
            }
        } else if (SINGLE_CHAR_SYMBOLS.count(std::string(1, current_char())) ||
                   ([&]() {
                       if (peek_next() == '\0') return false;
                       std::string sym;
                       sym += current_char();
                       sym += peek_next();
                       return DOUBLE_CHAR_SYMBOLS.count(sym) > 0;
                   })()) {
            tokens.push_back(read_symbol());
        } else if (std::isalpha(current_char()) || current_char() == '_') {
            tokens.push_back(read_word());
        } else if (std::isdigit(current_char())) {
            tokens.push_back(read_number());
        } else {
            std::string c = !is_at_end() ? std::string(1, current_char()) : "EOF";
            throw CompilerError(
                std::string("Tokenizer failed at char ") + std::to_string(position) +
                ": '" + c + "' (line " + std::to_string(line) +
                ", column " + std::to_string(col) + ")",
                "ERROR", file_path);
        }
    }
    return tokens;
}

std::string token_type_name(int token_type) {
    switch (token_type) {
        case T_NAME: return "NAME";
        case T_KEYWORD: return "KEYWORD";
        case T_FLOAT_LITERAL: return "FLOAT_LITERAL";
        case T_INT_LITERAL: return "INT_LITERAL";
        case T_SYMBOL: return "SYMBOL";
        case T_STRING: return "STRING";
        default: return "UNKNOWN";
    }
}

std::string escape_token_value(const std::string& value) {
    std::string escaped;
    for (char ch : value) {
        switch (ch) {
            case '\\': escaped += "\\\\"; break;
            case '\"': escaped += "\\\""; break;
            case '\n': escaped += "\\n"; break;
            case '\t': escaped += "\\t"; break;
            case '\r': escaped += "\\r"; break;
            default: escaped += ch; break;
        }
    }
    return escaped;
}

std::string format_tokens(const std::vector<Token>& tokens) {
    std::string out = "[";
    for (size_t i = 0; i < tokens.size(); i++) {
        if (i > 0) out += ", ";
        out += "<\"" + token_type_name(tokens[i].type) + "\", \"" + escape_token_value(tokens[i].value) + "\">";
    }
    out += "]";
    return out;
}
